File: tests/omp/declare_target5.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

  
  !! Omp.declare_target [cVarDef "Q"];
  !! Omp.declare_simd ~clause:[Uniform ["i"]; Linear (["k"],0); NotInbranch ] [cTopFunDef "P"];
  !! Omp.end_declare_target [cTopFunDef "accum"];

    let tg_loopi = [occIndex 1; cFor_c ""] in

  !! Omp.parallel_for ~clause:[Reduction (Plus, ["tmp"])] tg_loopi;
  !! Omp.target ~clause:[Map_c (ToFrom, ["tmp"])] tg_loopi;

    let tg_loopk = [occIndex 0; cFor_c ""] in
  !! Omp.parallel_for ~clause:[Reduction (Plus, ["tmp1"])] tg_loopk;

)

File: tests/omp/declare_target5.cpp
const int N = 10000;
const int M = 1024;
float Q[N][N];
float P(const int i, const int k)
{
  return Q[i][k] * Q[k][i];
}

float accum(void)
{
  float tmp = 0.0;
  for (int i=0; i < N; i++) {
    float tmp1 = 0.0;
    for (int k=0; k < M; k++) {
      tmp1 += P(i,k);
    }
    tmp += tmp1;
  }
  return tmp;
}
